File: practice/beginners_selection/golang/ABC049C.go
package main

import "fmt"
import "strings"

func f(S string) bool {
	if S == "dreamer" || S == "eraser" || S == "dream" || S == "erase" {
		return true
	}
	if len(S) > 7 {
		if strings.HasSuffix(S,"dreamer") {
			// fmt.Println(S[:len(S)-7])
			return f(S[:len(S)-7])
		}
	}
	if len(S) > 6 {
		if strings.HasSuffix(S,"eraser") {
			// fmt.Println(S[:len(S)-6])
			return f(S[:len(S)-6])
		}
	}
	if len(S) > 5 {
		if strings.HasSuffix(S,"dream") || strings.HasSuffix(S,"erase") {
			// fmt.Println(S[:len(S)-5])
			return f(S[:len(S)-5])
		}
	}
	return false
}

func main() {
	var S string
	fmt.Scan(&S) 
	if f(S) {
		fmt.Println("YES")
	} else {
		fmt.Println("NO")
	}
}

File: practice/beginners_selection/python3/ABC049C.py
import sys
sys.setrecursionlimit(1000)

def f(S):
    if S == 'dreamer' or S == 'eraser' or S == 'dream' or S == 'erase':
        return True
    if S.endswith('dreamer'):
        return f(S[:-7])
    if S.endswith('eraser'):
        return f(S[:-6])
    if S.endswith('dream') or S.endswith('erase'):
        return f(S[:-5])
    return False

def main():
    S = input()
    if f(S):
        print('YES')
    else:
        print('NO')

if __name__ == '__main__':
    main()
